File: Litvinov group/semester 3/Lazy/Lazy/MultiThreadLazy.cs
using System;

namespace Lazy
{
    /// <summary>
    /// Lazy calculation with the protection from mulrithreading
    /// </summary>
    /// <typeparam name="T">the type of the answer</typeparam>
    public class MultiThreadLazy<T> : ILazy<T>
    {
        private volatile bool calledBefore;
        private T result;
        private Func<T> supplier;
        private Object lockObject = new Object();

        /// <summary>
        /// Get the function
        /// </summary>
        /// <param name="supplier">function</param>
        public MultiThreadLazy(Func<T> supplier)
        {
            this.supplier = supplier;
        }

        /// <summary>
        /// If called for the first time - calculate and return the result
        /// and be shure, that only one thread will calculate and change calledBefore,
        /// else return already calculated result
        /// </summary>
        /// <returns>the result of the calculations</returns>
        public T Get()
        {
            if (!calledBefore)
            {
                lock (lockObject)
                {
                    if (!calledBefore)
                    {
                        result = supplier();
                        calledBefore = true;
                        supplier = null;
                    }
                }
            }

            return result;
        }
    }
}

File: Litvinov group/semester 4/08.HW/08.HW/Lazy/MultiThreadLazy.fs
module MultiThreadLazy

open ILazy
open System

/// Lazy calculation with the protection from mulrithreading
type MultiThreadLazy<'a> (supplier : unit -> 'a) =
    [<VolatileField>]
    let mutable calledBefore = false
    let mutable result = None
    let lockObject = new Object()

    /// If called for the first time - calculate and return the result
    /// and be sure, that only one thread will calculate and change calledBefore,
    /// else return already calculated result
    interface ILazy<'a> with
        member this.Get () = 
            if not calledBefore
            then lock lockObject (fun() -> if not calledBefore
                                           then result <- Some(supplier())
                                                calledBefore <- true)
            result.Value
